src/docfx/validation/CustomRuleProvider.cs

```diff
@@ -0,0 +1,255 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.Docs.Validation;
+
+namespace Microsoft.Docs.Build
+{
+    internal class CustomRuleProvider
+    {
+        private readonly DocumentProvider _documentProvider;
+        private readonly Lazy<PublishUrlMap> _publishUrlMap;
+        private readonly MonikerProvider _monikerProvider;
+        private readonly ErrorBuilder _errorLog;
+        private readonly Config _config;
+        private readonly Dictionary<string, List<CustomRule>> _customRules = new Dictionary<string, List<CustomRule>>();
+
+        public CustomRuleProvider(
+            Config config,
+            FileResolver fileResolver,
+            DocumentProvider documentProvider,
+            Lazy<PublishUrlMap> publishUrlMap,
+            MonikerProvider monikerProvider,
+            ErrorBuilder errorLog)
+        {
+            _documentProvider = documentProvider;
+            _publishUrlMap = publishUrlMap;
+            _monikerProvider = monikerProvider;
+            _errorLog = errorLog;
+            _config = config;
+
+            var contentValidationRules = GetContentValidationRules(config, fileResolver);
+            var buildValidationRules = GetBuildValidationRules(config, fileResolver);
+            _customRules = MergeCustomRules(contentValidationRules, buildValidationRules);
+        }
+
+        public bool IsEnable(FilePath filePath, CustomRule customRule, string? moniker = null)
+        {
+            var canonicalVersion = _publishUrlMap.Value.GetCanonicalVersion(filePath);
+
+            // If content versioning not enabled for this depot, canonicalVersion will be null, content will always be the canonical version;
+            // If content versioning enabled and moniker is null, we should check file-level monikers to be sure;
+            // If content versioning enabled and moniker is not null, just compare canonicalVersion and moniker.
+            var isCanonicalVersion = string.IsNullOrEmpty(canonicalVersion) ? true :
+                string.IsNullOrEmpty(moniker) ? _monikerProvider.GetFileLevelMonikers(_errorLog, filePath).IsCanonicalVersion(canonicalVersion) :
+                canonicalVersion == moniker;
+
+            if (customRule.CanonicalVersionOnly && !isCanonicalVersion)
+            {
+                return false;
+            }
+
+            var pageType = _documentProvider.GetPageType(filePath);
+
+            return customRule.ContentTypes is null || customRule.ContentTypes.Contains(pageType);
+        }
+
+        public Error ApplyCustomRule(Error error)
+        {
+            if (TryGetCustomRule(error, out var customRule))
+            {
+                error = WithCustomRule(error, customRule);
+            }
+            return error;
+        }
+
+        public static Error WithCustomRule(Error error, CustomRule customRule, bool? enabled = null)
```

Comment:
Shall we also call it ApplyCustomRule?

Reply:
this method accepts `CustomRule customRule` so we still use old name `WithCustomRule`, I think.